main.go

```diff
@@ -101,19 +123,64 @@ func setThreads() {
 	}
 }
 
-func unixAddr(uri string) (string, error) {
+func socketAddr(uri string) (socketAddress, error) {
 	if uri == "" {
-		return "", errors.New("empty uri")
+		return socketAddress{}, errors.New("empty uri")
 
 	}
 	addr, err := url.Parse(uri)
 	if err != nil {
-		return "", err
+		return socketAddress{}, err
+	}
+
+	switch addr.Scheme {
+	case "", client.UnixSocketScheme:
+		return socketAddress{
+			scheme: client.UnixSocketScheme,
+			path:   addr.Host + addr.Path,
+			port:   0,
+		}, nil
+	case client.HybridVSockScheme:
+		hvsocket := strings.Split(addr.Path, ":")
+		// expected path:port
+		if len(hvsocket) != 2 {
+			return socketAddress{}, fmt.Errorf("Invalid hybrid vsock scheme: %s", uri)
+		}
+
+		var port uint64
+		if port, err = strconv.ParseUint(hvsocket[1], 10, 32); err != nil {
+			return socketAddress{}, fmt.Errorf("Invalid hybrid vsock port %s: %v", uri, err)
+		}
+
+		return socketAddress{
+			scheme: client.HybridVSockScheme,
+			path:   hvsocket[0],
+			port:   uint32(port),
+		}, nil
+	default:
+		return socketAddress{}, errors.New("invalid address scheme")
 	}
-	if addr.Scheme != "" && addr.Scheme != "unix" {
-		return "", errors.New("invalid address scheme")
+}
+
+func socketDial(addr socketAddress) (net.Conn, error) {
+	switch addr.scheme {
+	case client.UnixSocketScheme:
+		return net.Dial("unix", addr.path)
+	case client.HybridVSockScheme:
+		for i := 0; i < hybridVSockMaxConnectTries; i++ {
+			conn, err := client.HybridVSockDialer(addr.string(), time.Second)
+			if err == nil {
+				c := conn.(*net.UnixConn)
+				if f, err := c.File(); err == nil && f != nil {
+					return conn, nil
+				}
+			}
+			time.Sleep(hybridVSockConnectDelay)
+		}
+		return nil, errors.New("unable to connect hybrid vsock")
```

Comment:
How about adding the original error(s) to this error to help debug why the dial is failing?

Reply:
yeah, sounds good